### PR TITLE
Add two more options to dense_tiles

### PR DESCRIPTION
**Pull request by Nakaner**
* `-a` to output all tiles at the requested zoom level (more user friendly than -m) because you do not have to manually enter the result of `2^(2*zoom_level)`.
* `-M <count>` to output only the tiles which contain more than `<count>` nodes

I added both features as part of my master thesis because I needed this little tool. I would like to share them with other people.

### CODE REVIEW

File: dense_tiles/dense_tiles.cpp [hunk 1]
```diff
@@ -25,31 +25,42 @@ void print_help(const char* progname) {
...
-    std::cerr << "   --max <n> | -m <n>    list n densest tiles [100000]\n";
-    std::cerr << "   --single | -s         compute for single tiles, not meta tiles\n";
-    std::cerr << "   --progress | -p       display progress bar\n";
+    std::cerr << "   --help | -h              this help\n";
+    std::cerr << "   --zoom <n> | -z <n>      compute for zoom n [14]\n";
+    std::cerr << "   --max <n> | -m <n>       list n densest tiles [100000]\n";
+    std::cerr << "   --min_nodes <n> | -M <n> list all tiles with more than n nodes;\n";
+    std::cerr << "                            only relevant if your --max setting would list\n";
+    std::cerr << "                            tiles with less than n nodes\n";
+    std::cerr << "   --all-tiles | -a         list all tiles (overwrites -m)\n";
```
Comment (joto) [inline dense_tiles/dense_tiles.cpp]: How about using `-m 0` for this instead of `-a`? This would remove the possible confusion and checks when both options are used.

File: dense_tiles/dense_tiles.cpp [hunk 2]
```diff
@@ -25,31 +25,42 @@ void print_help(const char* progname) {
...
     std::cerr << "left tile in an 8x8 tile block will be output.\n";
-    std::cerr << "   --help | -h           this help\n";
-    std::cerr << "   --zoom <n> | -z <n>   compute for zoom n [14]\n";
-    std::cerr << "   --max <n> | -m <n>    list n densest tiles [100000]\n";
-    std::cerr << "   --single | -s         compute for single tiles, not meta tiles\n";
-    std::cerr << "   --progress | -p       display progress bar\n";
+    std::cerr << "   --help | -h              this help\n";
+    std::cerr << "   --zoom <n> | -z <n>      compute for zoom n [14]\n";
+    std::cerr << "   --max <n> | -m <n>       list n densest tiles [100000]\n";
+    std::cerr << "   --min_nodes <n> | -M <n> list all tiles with more than n nodes;\n";
```
Comment (joto) [inline dense_tiles/dense_tiles.cpp]: The option should be called `--min-nodes` with a dash instead of an underscore.

File: dense_tiles/dense_tiles.cpp [hunk 3]
```diff
@@ -139,6 +159,8 @@ int main(int argc, char* argv[]) {
     for (unsigned int i = 0; i < grid.size(); ++i) {
         if (grid[i]) {
             sorter.emplace_back(grid[i], i);
+        } else {
+            sorter.emplace_back(0, i);
```
Comment (joto) [inline dense_tiles/dense_tiles.cpp]: Why is this here?

**Review by joto — CHANGES_REQUESTED** [review]
See the comments in the code.

File: dense_tiles/dense_tiles.cpp (hunk 3 above)
Comment (Nakaner) [inline dense_tiles/dense_tiles.cpp]: If the else clause missed, tiles without any nodes would not be printed. I had a use case in my master thesis where it is useful to have a list of all tiles—even the empty ones.

I will add a comment in the source code which explains it.

I will set the default value of `min_nodes` to 1 in order to keep the current behaviour and be backward compatible.

Comment (Nakaner) [inline dense_tiles/dense_tiles.cpp]: If the else clause missed, tiles without any nodes would not be printed. I had a use case in my master thesis where it is useful to have a list of all tiles—even the empty ones.

EDIT:  Thank you for your comment, these lines currently look really stupid. The else statement could be replace by `else if (min_nodes == 0)` which keeps the program as quick as before for most uses cases (except printing out all tiles).

I will add a comment in the source code which explains it.

I will set the default value of min_nodes to 1 in order to keep the current behaviour and be backward compatible.

**Comment by Nakaner** [comment]
I pushed a new commit which should fix everything.

File: dense_tiles/dense_tiles.cpp [hunk 4]
```diff
@@ -25,31 +25,46 @@ void print_help(const char* progname) {
...
+       { "max",       required_argument, 0, 'm' },
+       { "min-nodes", required_argument, 0, 'M' },
+       { "single",    no_argument,       0, 's' },
+       { "count",     no_argument,       0, 'c' },
+       { "progress",  no_argument,       0, 'p' },
        { 0, 0, 0, 0 } };
 
     while (true) {
-        const int c = getopt_long(argc, argv, "hm:psz:", long_options, 0);
+        const int c = getopt_long(argc, argv, "hm:M:apcsz:", long_options, 0);
```
Comment (joto) [inline dense_tiles/dense_tiles.cpp]: Still has 'a' in here.

File: dense_tiles/dense_tiles.cpp [hunk 5]
```diff
@@ -139,6 +165,10 @@ int main(int argc, char* argv[]) {
     for (unsigned int i = 0; i < grid.size(); ++i) {
         if (grid[i]) {
             sorter.emplace_back(grid[i], i);
+        } else if (min_nodes == 0) {
+            // This will add all tiles which contain no nodes. It is useful
+            // if you want a list of all tiles an a list how "large" each tile is.
+            sorter.emplace_back(0, i);
         }
```
Comment (joto) [inline dense_tiles/dense_tiles.cpp]: `if (grid[i] >= min_nodes) {` ? That would also get rid of the if-condition in new line 185-188.

File: dense_tiles/dense_tiles.cpp [hunk 6]
```diff
@@ -25,31 +25,46 @@ void print_help(const char* progname) {
...
     std::cerr << "left tile in an 8x8 tile block will be output.\n";
-    std::cerr << "   --help | -h           this help\n";
-    std::cerr << "   --zoom <n> | -z <n>   compute for zoom n [14]\n";
-    std::cerr << "   --max <n> | -m <n>    list n densest tiles [100000]\n";
-    std::cerr << "   --single | -s         compute for single tiles, not meta tiles\n";
-    std::cerr << "   --progress | -p       display progress bar\n";
+    std::cerr << "   --help | -h              this help\n";
+    std::cerr << "   --zoom <n> | -z <n>      compute for zoom n [14]\n";
+    std::cerr << "   --max <n> | -m <n>       list n densest tiles [100000]\n";
+    std::cerr << "   --min-nodes <n> | -M <n> list all tiles with more than n nodes;\n";
```
Comment (joto) [inline dense_tiles/dense_tiles.cpp]: `with more than n nodes [1]` (add ` [1]`, the default, remove `;`)

File: dense_tiles/dense_tiles.cpp [hunk 7]
```diff
@@ -25,31 +25,46 @@ void print_help(const char* progname) {
...
+    std::cerr << "   --progress | -p          display progress bar\n";
+    std::cerr << "\n";
+    std::cerr << "Conflict of --max` vs. `--min-nodes` setting:\n";
+    std::cerr << "  In case of a conflict between these two settings, the setting which\n";
+    std::cerr << "  sets the lower number of printed tiles will stop the output of further\n";
+    std::cerr << "  tiles.\n";
+    std::cerr << "  Use `--max 0` (it's a shortcut) to get all tiles printed which\n";
+    std::cerr << "  contain more or equal than `--min-nodes`. Use the `--min-nodes 0` to\n";
+    std::cerr << "  print exactly that number of tiles you requested using the `--max`\n";
+    std::cerr << "  setting (except there are not that much tiles at the zoom level. ;-)\n";
```
Comment (joto) [inline dense_tiles/dense_tiles.cpp]: Not sure I understand that explanation. :-) Shouldn't we simply change the description above to "List the (meta) tiles  in the input file ordered by node density, highest first". And the description of `--max` to `list a maximum of n tiles...`

**Review by joto — CHANGES_REQUESTED** [review]
Much better!

File: dense_tiles/dense_tiles.cpp (hunk 7 above)
Comment (Nakaner) [inline dense_tiles/dense_tiles.cpp]: Yes, "List the (meta) tiles in the input file ordered by node density, highest first" is a good description. I adopted it.

If you use `-z 12 -s --max=10 --min-nodes=5` and a large input file, dense-tiles will only list the ten largest tiles at zoom level 12. All other tiles with at least 5 nodes will not be printed. If you use `-z 12 -s --max=16777216 --min-nodes=5`, all tiles with at least 5 nodes will be printed because `--min-nodes` sets the limit which is reached first.

I thought a lot how to describe this behaviour. Your questions shows that I did not think enough ;-) or the feature should be dropped. I suggest to postpone this discussion to the hack weekend next weekend.

File: dense_tiles/dense_tiles.cpp (hunk 6 above)
Comment (Nakaner) [inline dense_tiles/dense_tiles.cpp]: good suggestion, adopted, will be part of the next commit

File: dense_tiles/dense_tiles.cpp (hunk 5 above)
Comment (Nakaner) [inline dense_tiles/dense_tiles.cpp]: good suggestion, adopted, will be part of the next commit

File: dense_tiles/dense_tiles.cpp (hunk 7 above)
Comment (joto) [inline dense_tiles/dense_tiles.cpp]: Okay :-)
